web/pandas/pdeps/0013-logical-type-system.md

```diff
@@ -0,0 +1,161 @@
+| string       | N/A     | "string"         | N/A                      |
+| datetime     | N/A     | "datetime64[us]" | "timestamp[us][pyarrow]" |
+
+"string[pyarrow]" is excluded from the above table because it is misleading; while "int64[pyarrow]" definitely gives you a pyarrow backed string, "string[pyarrow]" gives you a pandas extension array which itself then uses pyarrow, which can introduce behavior differences (see [issue 58321](https://github.com/pandas-dev/pandas/issues/58321)).
+
+If you wanted to try and be more explicit about using pyarrow, you could use the ``pd.ArrowDtype`` wrapper. But this unfortunately exposes gaps when trying to use that pattern across all backends:
+
+| logical type | NumPy    | pandas extension | pyarrow                           |
+| int          | np.int64 | pd.Int64Dtype()  | pd.ArrowDtype(pa.int64())         |
+| string       | N/A      | pd.StringDtype() | pd.ArrowDtype(pa.string())        |
+| datetime     | N/A      | ???              | pd.ArrowDtype(pa.timestamp("us")) |
+
+It would stand to reason in this approach that you could use a ``pd.DatetimeDtype()`` but no such type exists (there is a ``pd.DatetimeTZDtype`` which requires a timezone).
+
+The third issue is that the extent to which pandas may support any given type is unclear. Issue [#58307](https://github.com/pandas-dev/pandas/issues/58307) highlights one example. It would stand to reason that you could interchangeably use a pandas datetime64 and a pyarrow timestamp, but that is not always true. Another common example is the use of NumPy fixed length strings, which users commonly try to use even though we claim no real support for them (see [#5764](https://github.com/pandas-dev/pandas/issues/57645)).
+
+## Assessing the Current Type System(s)
+
+A best effort at visualizing the current type system(s) with types that we currently "support" or reasonably may want to is shown [in this comment](https://github.com/pandas-dev/pandas/issues/58141#issuecomment-2047763186). Note that this does not include the ["pyarrow_numpy"](https://github.com/pandas-dev/pandas/pull/58451) string data type or the string data type that uses the NumPy 2.0 variable length string data type (see [comment](https://github.com/pandas-dev/pandas/issues/57073#issuecomment-2080798081)) as they are under active discussion.
+
+## Proposal
+
+Derived from the hierarchical visual in the previous section, this PDEP proposes that pandas supports at least all of the following _logical_ types, excluding any type widths for brevity:
+
+  - Signed Integer
+  - Unsigned Integer
+  - Floating Point
+  - Fixed Point
+  - Boolean
+  - Date
+  - Datetime
+  - Duration
+  - Interval
+  - Period
+  - Binary
+  - String
+  - Dictionary
+  - List
+  - Struct
+
+To ensure we maintain all of the current functionality of our existing type system(s), a base type structure would need to look something like:
+
+```python
+class BaseType:
```

Comment:
would this then get mixed into ExtensionDtype?  if not, what is this used for?

Reply:
I suppose it depends on whether we as a team want to go down the path of offering constructors like `pd.string()` or `pd.StringDtype()`; with the former approach this would not mix into the ExtensionDtype, since that would be closer to a physical type. If we went with the latter approach, we would be repurposing / clarifying the scope of the pandas extension type system to be totally logical

Reply:
As I think through it more it may make sense for these to be separate; the existing ExtensionDtype refers to a structure that pandas wants to contain a particular set of values. For extension types it is possible for there not to be a logical type representation in pandas

Maybe this should be renamed to `BaseLogicalType` to clairfy

Reply:
then im unclear on what we're doing with it?  and if EA authors are supposed to implement something for it

Reply:
That's a good point - for third party EAs we would not be able to offer a logical type, since we don't know what the EA is. Let's see what other teammates think

Reply:
That might be a good reason to prefer the factory function constructors in the PDEP, so that we aren't confusing these logical types with 3rd party implemented extensions

Reply:
In my mind this is the abstact interface that is essentially some `BaseDtype`, and the `ExtensionDtype` could inherit from that adding some helpers for external implemenations (or if that is not needed, those two could be exactly the sane)?

I don't really understand the point about EAs not being logical. Of course they don't provide one of the logical types that are listed in this PDEP, which is exactly their point to _extend_ the builtin types. But in my mind the EAs have always already been logical types, because they wrap an existing array giving a different meaning to it (like how it originally started, storing bytes interpreting it as UUID)

Reply:
Similar to my comment https://github.com/pandas-dev/pandas/pull/58455#discussion_r1582376637 the next revision of this PDEP will have to add terminology to clarify what is considered logical versus not.

I also imagine the theoretical, highly unlikely case where two third party EAs try to compete in the same space. Like if someone decided they wanted to revive CyberPandas and created a new library called IPPandas, a user would then be able to pick either the CyberPandas implementation or the IPPandas implementation. I would consider both of those third party libraries _physical_ insofar that they have control over how they choose to layout their data buffers / masks and how they choose to implement algorithms. But the logical type for those would be an "IP Type"

Reply:
While speaking about physical and logical types here for the discussion is useful, another question is what would we actually provide in APIs? What is the `.dtype` that users interact with? That's the logical type? 

In theory we could implement _only_ logical types in pandas, and they can have different instantiations representing the logical type backed by a different physical layout (i.e. having a different `.physical_type`)

For external extension dtypes, I think it will be very hard to extend this concept of logical types to multiple external libraries, like your example of two independent external libraries implementing the same logical type. How would those two interact, or know from each other they are the same logical type? Personally, I would just consider every external EA its _own_ logical type (yes, that might mean that there are two libraries providing a very similar logical type, but so be it)

Reply:
> What is the `.dtype` that users interact with? That's the logical type?

Yea that's where my head is at - essentially they would see / interact with a subclass of the `BaseType` shown here



> For external extension dtypes, I think it will be very hard to extend this concept of logical types to multiple external libraries, like your example of two independent external libraries implementing the same logical type.

Yea this is very tricky. I don't know what the best answer is but I don't really want to muddy the waters between an Extension type and a Logical type; I think these are two totally different things. But this PDEP would need to clarify further how users would use Extension types knowing there is no equivalent logical type for them (or maybe we need to have them register themselves as a logical type?)

Reply:
> I don't really want to muddy the waters between an Extension type and a Logical type; I think these are two totally different things

I still don't fully follow why they are totally different things. I know our _own_ types are messy (with the mixture of numpy, numpy-based extension dtypes and pyarrow extension dtypes), but for the _external_ extension dtypes that I know, those are almost all to be considered as "logical" types (eg geometries, tensors, ..)

Reply:
Because our third party extension arrays give EA authors theoretical control over data buffers, masks, and algorithms. A logical type should have none of these

Reply:
> Because our third party extension arrays give EA authors theoretical control over data buffers, masks, and algorithms. A logical type should have none of these

That's definitely true that in that sense an EA is also a physical type. But at the same time, in the idea that a _user_ generally specifies the logical type, all external EAs are in practice also a logical type. 

So I understand that those are different concepts, but I think in _practice_ the external EA serves both purposes, and so what I don't fully see is 1) how it would be possible for external EAs to distinguish those concepts, and 2) what the value would be of doing so.